### PR TITLE
rgw/multisite: avoid writing multipart parts to the bucket index log

### PR DESCRIPTION
**Pull request by jzhu116-bloomberg**
Fixes: https://tracker.ceph.com/issues/21800
Signed-off-by: Jane Zhu [jzhu116@bloomberg.net](mailto:jzhu116@bloomberg.net)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/driver/rados/rgw_sal_rados.cc [hunk 1]
```diff
@@ -2904,7 +2904,8 @@ int RadosMultipartWriter::complete(size_t accounted_size, const std::string& eta
                        uint32_t flags)
 {
   return processor.complete(accounted_size, etag, mtime, set_mtime, attrs, delete_at,
-			    if_match, if_nomatch, user_data, zones_trace, canceled, rctx, flags);
+                            if_match, if_nomatch, user_data, zones_trace, canceled, rctx,
+                            flags & ~FLAG_LOG_OP);
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_sal_rados.cc:2908]: it makes sense to remove that flag for `MultipartWriter`s, and `RadosMultipartWriter` is a convenient place to do that, but it makes this change unnecessarily rados-specific. someday we'll support multisite sync for other sal backends too, and the behavior around multipart uploads will likely stay backend-agnostic

to make this more generic, `RGWPutObj::execute()` chould decide whether or not to pass FLAG_LOG_OP:
```diff
diff --git a/src/rgw/rgw_op.cc b/src/rgw/rgw_op.cc
index d1d332b0016..f5ed1089bfa 100644
--- a/src/rgw/rgw_op.cc
+++ b/src/rgw/rgw_op.cc
@@ -4587,12 +4587,16 @@ void RGWPutObj::execute(optional_yield y)
     emplace_attr(RGW_ATTR_OBJECT_RETENTION, std::move(obj_retention_bl));
   }
 
+  // don't track the individual parts of multipart uploads. they replicate in
+  // full after CompleteMultipart
+  const uint32_t complete_flags = multipart ? 0 : rgw::sal::FLAG_LOG_OP;
+
   tracepoint(rgw_op, processor_complete_enter, s->req_id.c_str());
   const req_context rctx{this, s->yield, s->trace.get()};
   op_ret = processor->complete(s->obj_size, etag, &mtime, real_time(), attrs,
                                (delete_at ? *delete_at : real_time()), if_match, if_nomatch,
                                (user_data.empty() ? nullptr : &user_data), nullptr, nullptr,
-                               rctx, rgw::sal::FLAG_LOG_OP);
+                               rctx, complete_flags);
   tracepoint(rgw_op, processor_complete_exit, s->req_id.c_str());
   if (op_ret < 0) {
     return;
```

**Review by cbodley — COMMENTED** [review]
looks like this covers all the writes/deletes for multipart parts and the multipart-meta entries 👍

File: src/cls/rgw/cls_rgw.cc [hunk 2]
```diff
@@ -1210,7 +1210,7 @@ int rgw_bucket_complete_op(cls_method_context_t hctx, bufferlist *in, bufferlist
     CLS_LOG_BITX(bitx_inst, 20,
 		 "INFO: %s: completing object remove key=%s",
 		 __func__, escape_str(remove_key.to_string()).c_str());
-    rc = complete_remove_obj(hctx, header, remove_key, default_log_op);
+    rc = complete_remove_obj(hctx, header, remove_key, default_log_op && op.log_remove_objs);
```
[the remark this answers is not in the document]
Comment (jzhu116-bloomberg) [reply src/cls/rgw/cls_rgw.cc]: No I don't see other places where it's being used. I originally thought to make rgw_rados more generic. But it's probably not a good idea to bloat the code unnecessarily. We can always make it more generic if more use cases coming up. For now, I removed the extra flag.

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-02-02_18:51:03-rgw-wip-cbodley2-testing-distro-default-smithi/

**Pull request merged**